File: internal/protocol/wire.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"io"
)

// Magic bytes identify maboo-wire protocol frames.
var Magic = [2]byte{0x4D, 0x42} // "MB"

// Version is the current protocol version.
const Version uint8 = 0x01

// FrameHeaderSize is the fixed size of a frame header in bytes.
const FrameHeaderSize = 14

// Message types define the purpose of each frame.
const (
	TypeRequest     uint8 = 0x01 // Go → PHP: new HTTP request
	TypeResponse    uint8 = 0x02 // PHP → Go: HTTP response
	TypeStreamData  uint8 = 0x03 // Bidirectional: WebSocket frame
	TypeStreamClose uint8 = 0x04 // Either: close WebSocket connection
	TypeWorkerReady uint8 = 0x05 // PHP → Go: worker is available
	TypeWorkerStop  uint8 = 0x06 // Go → PHP: graceful shutdown
	TypePing        uint8 = 0x07 // Health check (ping/pong)
	TypeError       uint8 = 0x08 // Error reporting
)

// Flags modify frame behavior.
const (
	FlagCompressed uint8 = 1 << 0 // Payload is compressed
	FlagChunked    uint8 = 1 << 1 // Chunked transfer
	FlagFinal      uint8 = 1 << 2 // Final chunk in sequence
)

// Frame represents a single maboo-wire protocol frame.
type Frame struct {
	Type     uint8
	Flags    uint8
	StreamID uint16
	Headers  []byte // msgpack encoded
	Payload  []byte // raw bytes
}

// WriteFrame encodes and writes a frame to the given writer.
func WriteFrame(w io.Writer, f *Frame) error {
	header := make([]byte, FrameHeaderSize)
	header[0] = Magic[0]
	header[1] = Magic[1]
	header[2] = Version
	header[3] = f.Type
	header[4] = f.Flags
	binary.BigEndian.PutUint16(header[5:7], f.StreamID)

	// Header size as 3 bytes (big-endian uint24)
	hdrSize := len(f.Headers)
	header[7] = byte(hdrSize >> 16)
	header[8] = byte(hdrSize >> 8)
	header[9] = byte(hdrSize)

	binary.BigEndian.PutUint32(header[10:14], uint32(len(f.Payload)))

	if _, err := w.Write(header); err != nil {
		return fmt.Errorf("writing frame header: %w", err)
	}
	if len(f.Headers) > 0 {
		if _, err := w.Write(f.Headers); err != nil {
			return fmt.Errorf("writing frame headers: %w", err)
		}
	}
	if len(f.Payload) > 0 {
		if _, err := w.Write(f.Payload); err != nil {
			return fmt.Errorf("writing frame payload: %w", err)
		}
	}
	return nil
}

// ReadFrame reads and decodes a frame from the given reader.
func ReadFrame(r io.Reader) (*Frame, error) {
	header := make([]byte, FrameHeaderSize)
	if _, err := io.ReadFull(r, header); err != nil {
		return nil, fmt.Errorf("reading frame header: %w", err)
	}

	if header[0] != Magic[0] || header[1] != Magic[1] {
		return nil, fmt.Errorf("invalid magic bytes: 0x%02x%02x", header[0], header[1])
	}
	if header[2] != Version {
		return nil, fmt.Errorf("unsupported protocol version: %d", header[2])
	}

	f := &Frame{
		Type:     header[3],
		Flags:    header[4],
		StreamID: binary.BigEndian.Uint16(header[5:7]),
	}

	hdrSize := int(header[7])<<16 | int(header[8])<<8 | int(header[9])
	payloadSize := binary.BigEndian.Uint32(header[10:14])

	if hdrSize > 0 {
		f.Headers = make([]byte, hdrSize)
		if _, err := io.ReadFull(r, f.Headers); err != nil {
			return nil, fmt.Errorf("reading frame headers (%d bytes): %w", hdrSize, err)
		}
	}
	if payloadSize > 0 {
		f.Payload = make([]byte, payloadSize)
		if _, err := io.ReadFull(r, f.Payload); err != nil {
			return nil, fmt.Errorf("reading frame payload (%d bytes): %w", payloadSize, err)
		}
	}

	return f, nil
}

// NewPingFrame creates a PING health check frame.
func NewPingFrame() *Frame {
	return &Frame{Type: TypePing, Payload: []byte("ping")}
}

// NewPongFrame creates a PONG response frame.
func NewPongFrame() *Frame {
	return &Frame{Type: TypePing, Payload: []byte("pong")}
}

// NewWorkerReadyFrame creates a WORKER_READY signal frame.
func NewWorkerReadyFrame() *Frame {
	return &Frame{Type: TypeWorkerReady}
}

// NewWorkerStopFrame creates a WORKER_STOP signal frame.
func NewWorkerStopFrame() *Frame {
	return &Frame{Type: TypeWorkerStop}
}

// NewErrorFrame creates an ERROR frame with a message.
func NewErrorFrame(msg string) *Frame {
	return &Frame{Type: TypeError, Payload: []byte(msg)}
}

File: php-sdk/src/Protocol/Wire.php
<?php

declare(strict_types=1);

namespace Maboo\Protocol;

class Wire
{
    private const MAGIC = "\x4D\x42"; // "MB"
    private const VERSION = 0x01;
    private const HEADER_SIZE = 14;

    // Message types (must match Go constants)
    public const TYPE_REQUEST = 0x01;
    public const TYPE_RESPONSE = 0x02;
    public const TYPE_STREAM_DATA = 0x03;
    public const TYPE_STREAM_CLOSE = 0x04;
    public const TYPE_WORKER_READY = 0x05;
    public const TYPE_WORKER_STOP = 0x06;
    public const TYPE_PING = 0x07;
    public const TYPE_ERROR = 0x08;

    // Flags
    public const FLAG_COMPRESSED = 0x01;
    public const FLAG_CHUNKED = 0x02;
    public const FLAG_FINAL = 0x04;

    /**
     * Read a frame from the given stream (default: STDIN).
     */
    public static function readFrame($stream = null): Frame
    {
        $stream = $stream ?? STDIN;

        $header = self::readExact($stream, self::HEADER_SIZE);

        // Validate magic
        if ($header[0] !== self::MAGIC[0] || $header[1] !== self::MAGIC[1]) {
            throw new \RuntimeException(sprintf(
                'Invalid magic bytes: 0x%02x%02x',
                ord($header[0]),
                ord($header[1])
            ));
        }

        // Validate version
        if (ord($header[2]) !== self::VERSION) {
            throw new \RuntimeException('Unsupported protocol version: ' . ord($header[2]));
        }

        $type = ord($header[3]);
        $flags = ord($header[4]);
        $streamId = unpack('n', substr($header, 5, 2))[1]; // big-endian uint16

        // Header size as 3 bytes (big-endian uint24)
        $hdrSize = (ord($header[7]) << 16) | (ord($header[8]) << 8) | ord($header[9]);

        // Payload size as 4 bytes (big-endian uint32)
        $payloadSize = unpack('N', substr($header, 10, 4))[1];

        $headers = '';
        if ($hdrSize > 0) {
            $headers = self::readExact($stream, $hdrSize);
        }

        $payload = '';
        if ($payloadSize > 0) {
            $payload = self::readExact($stream, $payloadSize);
        }

        return new Frame($type, $flags, $streamId, $headers, $payload);
    }

    /**
     * Write a frame to the given stream (default: STDOUT).
     */
    public static function writeFrame(Frame $frame, $stream = null): void
    {
        $stream = $stream ?? STDOUT;

        $hdrSize = strlen($frame->headers);
        $payloadSize = strlen($frame->payload);

        $header = self::MAGIC;
        $header .= chr(self::VERSION);
        $header .= chr($frame->type);
        $header .= chr($frame->flags);
        $header .= pack('n', $frame->streamId); // big-endian uint16

        // Header size as 3 bytes (big-endian uint24)
        $header .= chr(($hdrSize >> 16) & 0xFF);
        $header .= chr(($hdrSize >> 8) & 0xFF);
        $header .= chr($hdrSize & 0xFF);

        $header .= pack('N', $payloadSize); // big-endian uint32

        fwrite($stream, $header);
        if ($hdrSize > 0) {
            fwrite($stream, $frame->headers);
        }
        if ($payloadSize > 0) {
            fwrite($stream, $frame->payload);
        }
        fflush($stream);
    }

    /**
     * Read exact number of bytes from stream.
     */
    private static function readExact($stream, int $length): string
    {
        $data = '';
        $remaining = $length;
        while ($remaining > 0) {
            $chunk = fread($stream, $remaining);
            if ($chunk === false || $chunk === '') {
                throw new \RuntimeException("Stream closed: expected $length bytes, got " . strlen($data));
            }
            $data .= $chunk;
            $remaining -= strlen($chunk);
        }
        return $data;
    }
}
